Mathlib/Topology/MetricSpace/UniformConvergence.lean

```diff
@@ -0,0 +1,280 @@
+noncomputable instance : PseudoEMetricSpace (α →ᵤ β) where
+  edist f g := ⨆ x, edist (f x) (g x)
+  edist_self := by simp
+  edist_comm := by simp [edist_comm]
+  edist_triangle f₁ f₂ f₃ := calc
+    ⨆ x, edist (f₁ x) (f₃ x) ≤ ⨆ x, edist (f₁ x) (f₂ x) + edist (f₂ x) (f₃ x) :=
+      iSup_mono fun _ ↦ edist_triangle _ _ _
+    _ ≤ (⨆ x, edist (f₁ x) (f₂ x)) + (⨆ x, edist (f₂ x) (f₃ x)) := iSup_add_le _ _
+  toUniformSpace := inferInstance
+  uniformity_edist := by
+    rw [UniformFun.hasBasis_uniformity_of_basis α β uniformity_basis_edist |>.eq_biInf]
+    simp only [Function.comp_apply, UniformFun.gen, Set.mem_setOf_eq]
+    refine le_antisymm ?_ <| iInf₂_mono ?_
+    · refine iInf₂_mono' fun ε hε ↦ ?_
+      obtain ⟨δ, hδ, hδε⟩ := exists_between hε
+      exact ⟨δ, hδ, by simpa [iSup_lt_iff] using fun f g h ↦ ⟨δ, hδε, fun x ↦ (h x).le⟩⟩
+    · simpa using fun ε hε f g h x ↦ (le_iSup _ x).trans_lt h
+
+lemma edist_def (f g : α →ᵤ β) :
+    edist f g = ⨆ x, edist (toFun f x) (toFun g x) :=
+  rfl
+
+noncomputable instance {β : Type*} [EMetricSpace β] : EMetricSpace (α →ᵤ β) where
+  eq_of_edist_eq_zero {f g} h := funext fun x ↦ eq_of_edist_eq_zero <| le_antisymm
+    ((edist_def f g ▸ h) ▸ le_iSup (fun y ↦ edist (f y) (g y)) x) (zero_le _)
+
+lemma lipschitzWith_iff {f : γ → α →ᵤ β} {K : ℝ≥0} :
+    LipschitzWith K f ↔ ∀ c, LipschitzWith K (fun x ↦ toFun (f x) c) := by
+  simp [LipschitzWith, edist_def, forall_comm (α := α), toFun, ofFun]
+
+lemma lipschitzWith_ofFun_iff {f : γ → α → β} {K : ℝ≥0} :
+    LipschitzWith K (fun x ↦ ofFun (f x)) ↔ ∀ c, LipschitzWith K (f · c) :=
+  lipschitzWith_iff
+
+/-- If `f : α → γ → β` is a family of a functions, all of which are Lipschitz with the
+same constant, then the family is uniformly equicontinuous. -/
+lemma _root_.LipschitzWith.uniformEquicontinuous (f : α → γ → β) (K : ℝ≥0)
+    (h : ∀ c, LipschitzWith K (f c)) : UniformEquicontinuous f := by
+  rw [uniformEquicontinuous_iff_uniformContinuous]
+  rw [← lipschitzWith_ofFun_iff] at h
+  exact h.uniformContinuous
+
+lemma lipschitzOnWith_iff {f : γ → α →ᵤ β} {K : ℝ≥0} {s : Set γ} :
+    LipschitzOnWith K f s ↔ ∀ c, LipschitzOnWith K (fun x ↦ toFun (f x) c) s := by
+  simp [lipschitzOnWith_iff_restrict, lipschitzWith_iff]
+  rfl
+
+lemma lipschitzWith_eval (x : α) :
+    LipschitzWith 1 (fun f : α →ᵤ β ↦ f x) := by
+  intro f g
+  simpa [edist_def] using le_iSup (fun y ↦ edist (toFun f y) (toFun g y)) x
+
+end EMetric
+
+section Metric
+
+variable [PseudoMetricSpace β]
+
+noncomputable instance [BoundedSpace β] : PseudoMetricSpace (α →ᵤ β) :=
+  PseudoEMetricSpace.toPseudoMetricSpaceOfDist
+    (fun f g ↦ ⨆ x, dist (toFun f x) (toFun g x))
+    (fun _ _ ↦ by
+      have := BoundedSpace.bounded_univ (α := β) |>.ediam_ne_top.lt_top
+      refine (iSup_le fun x ↦ EMetric.edist_le_diam_of_mem ?_ ?_).trans_lt this |>.ne
+      all_goals trivial)
+    (fun _ _ ↦ by simp [edist_def, ENNReal.toReal_iSup (fun _ ↦ edist_ne_top _ _), dist_edist])
+
```

Comment:
Here I'd also suggest adding:

```suggestion

lemma dist_def [BoundedSpace β] (f g : α →ᵤ β) :
    dist f g = ⨆ x, dist (toFun f x) (toFun g x) :=
  rfl

lemma dist_le [BoundedSpace β] {f g : α →ᵤ β} {C : ℝ} (hC : 0 ≤ C) :
    dist f g ≤ C ↔ ∀ x, dist (toFun f x) (toFun g x) ≤ C := by
  simp_rw [dist_edist, ← ENNReal.le_ofReal_iff_toReal_le (edist_ne_top _ _) hC, edist_le]

```

(Especially since the second one is not that trivial)